checker/src/org/checkerframework/checker/regex/RegexQualifiedTransfer.java

```diff
@@ -45,82 +59,218 @@ public RegexQualifiedTransfer(QualAnalysis<QualParams<Regex>> analysis) {
-                QualParams<Regex> regex;
-                Node count = n.getArgument(1);
-                if (count instanceof IntegerLiteralNode) {
-                    IntegerLiteralNode iln = (IntegerLiteralNode) count;
-                    Integer groupCount = iln.getValue();
-                    regex = new QualParams<>(new GroundQual<Regex>(new RegexVal(groupCount)));
-                } else {
-                    regex = new QualParams<>(new GroundQual<Regex>(new RegexVal(0)));
-                }
-                QualValue<QualParams<Regex>> newResultValue = analysis
-                        .createSingleAnnotationValue(regex,
-                                result.getResultValue().getType().getUnderlyingType().getOriginalType());
-                return new RegularTransferResult<>(newResultValue,
-                        result.getRegularStore());
+        if (receiver instanceof ClassNameNode) {
+            ClassNameNode cnn = (ClassNameNode) receiver;
+            String receiverName = cnn.getElement().toString();
+            if (isRegexUtil(receiverName)) {
+                result = handleRegexUtil(n, method, result);
             }
         }
-
         return result;
     }
 
+    private TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> handleRegexUtil(
+            MethodInvocationNode n, ExecutableElement method,
+            TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> resultIn) {
+        if (ElementUtils.matchesElement(method, IS_REGEX_METHOD_NAME, String.class, int.class)) {
+            // RegexUtil.isRegex(s, groups) method
+            // (No special case is needed for isRegex(String) because of
+            // the annotation on that method's definition.)
+
+            QualStore<QualParams<Regex>> thenStore = resultIn.getRegularStore();
+            QualStore<QualParams<Regex>> elseStore = thenStore.copy();
+            ConditionalTransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> newResult =
+                    new ConditionalTransferResult<>(resultIn.getResultValue(), thenStore, elseStore);
+            FlowExpressionContext context = FlowExpressionParseUtil.buildFlowExprContextForUse(n,
+                    analysis.getContext());
+            Receiver firstParam;
+            try {
+                // TODO: is this the easiest way to do this?
+                firstParam = FlowExpressionParseUtil.parse("#1", context,
+                        analysis.getContext().getTypeFactory().getPath(n.getTree()));
+            } catch (FlowExpressionParseException e) {
+                firstParam = null;
+                assert false;
+            }
+
+            // add annotation with correct group count (if possible,
+            // regex annotation without count otherwise)
+            Node count = n.getArgument(1);
+            int groupCount;
+            if (count instanceof IntegerLiteralNode) {
```

Comment:
I don't think this should use IntegerLiteralNode.  Rather, it should take advantage of the Constant Value Checker and be applicable to any expression that the Constant Value Checker can determine the run-time value of.